### PR TITLE
[Relay][Types][Tests] Add type solver unit tests for unifying quantified funcs

### PR DESCRIPTION
**Pull request by slyubomirsky**
See issue #3946 . This PR adds more tests for new changes to unifying quantified function types from #3559 .

Unfortunately, one of these unit tests crashes because of a unification bug whose solution is not obvious. I have included the case described in #3946 here so we could make progress towards a fix here.

### CODE REVIEW

File: tests/python/relay/test_type_solver.py [hunk 1]
```diff
@@ -224,6 +224,49 @@ def test_backward_solving_after_child_update():
...
+    assert unified == ft2
+
+
+def test_unify_quantified_funcs_var_order():
+    solver = make_solver()
+    a, b, c = relay.TypeVar('a'), relay.TypeVar('b'), relay.TypeVar('c')
+
+    ft1 = relay.FuncType([a, relay.TupleType([b, c])], relay.TupleType([a, b, c]), [a, b, c])
+    ft2 = relay.FuncType([a, relay.TupleType([a, c])], relay.TupleType([a, a, c]), [a, c])
+    # unified = solver.Unify(ft1, ft2) # crashes here but it shouldn't
```
[the remark this answers is not in the document]
Comment (jroesch) [reply tests/python/relay/test_type_solver.py]: We should check before we hit the unifier and throw an error

[the remark this answers is not in the document]
Comment (slyubomirsky) [reply tests/python/relay/test_type_solver.py]: Check for what namely? Differing number of args? I believe @MarisaKirisame needed to allow for instantiating new args to make CPS work

[the remark this answers is not in the document]
Comment (MarisaKirisame) [reply tests/python/relay/test_type_solver.py]: I think we should just do generalization.

[the remark this answers is not in the document]
Comment (jroesch) [reply tests/python/relay/test_type_solver.py]: @MarisaKirisame can you clearly write up a design and what will change from the design today including examples?

**Pull request merged**